software/oada_upload/src/oada_upload.ts

```diff
@@ -0,0 +1,387 @@
+
+  console.log(`Connecting to OADA: ${domain}`);
+  const oada = await connect({ domain, token });
+  
+  // Infinite loop to continuously check the db for unsent data
+  for (;;) {
+    // Extract x number of unsent data points
+    const gps = await db.query(
+      `SELECT gps.time as time_tz, extract(epoch from gps.time) as time_epoch, gps.lat, gps.lng, gps_sent.sent
+        FROM gps
+        JOIN gps_sent ON gps.time = gps_sent.g_time
+
+        WHERE sent = FALSE
+        ORDER BY time DESC
+        LIMIT $1;`,
+      [batchsize]
+    );
+
+    // If we get no data from the databse, we sent everything.
+    // Wait 1s for the data to start to refill the db and requery
+    if (!gps.rows.length) {
+      console.log('No unsent data found in database');
+      await sleep(1000);
+      continue;
+    }
+    
+    console.info(`Found batch of ${gps.rows.length} GPS points`);
+    console.debug(`Head time: ${gps.rows[0].time_tz}`);
+
+    // Create DataIndex object(?) for filling data in
+    const data: DataIndex = {};
+    // For each datapoint used, extract data and begin to sort them into buckets to be uploaded
+    // Due to how the OADA server is setup, we the buckets will be by hour collected
+    gps.rows.forEach((p) => {
+      //console.debug(`Processing GPS time: ${p.time_tz}`);
+
+      // Extract time and use it to create path that it will be uploaded it
+      const t = moment.unix(p.time_epoch);
+      const day = t.format('YYYY-MM-DD');
+      const hour = t.format('HH');
+      const path = `/bookmarks/isoblue/device-index/${id}/location/day-index/${day}/hour-index/${hour}`;
+      // Create a UUID that can be coarsely sorted by time of creation
+      // TODO: find a way to use the timestamp to create the id instead of the current time
+      const pId = ksuid.randomSync().string;
+
+      // If the current bucket does not exist, create it
+      if (!data[path]) {
+        data[path] = {
+          epochs: [],
+          timetzs: [],
+          locations: {},
+        };
+      }
+
+      // Insert our data into the bucket
+      data[path].epochs.push(p.time_epoch);
+      data[path].timetzs.push(p.time_tz);
+      data[path].locations[pId] = {
+        id: pId,
+        time: {
+          value: p.time_epoch,
+        },
+        location: {
+          lat: p.lat,
+          lng: p.lng,
+        },
+      };
+      
+    });
+    // pEachSeries executes the following code for each bucket (path) that we created sequentially 
+    await pEachSeries(Object.keys(data), async (path) => {
+      var update_success = true;
+      var res;
+
+      // Attempt the put to OADA
+      try {
+        //console.debug(`GPS ID ${data[path].epochs.join(',')} to OADA ${path}`);
+        console.debug(`${data[path].epochs.length} points to OADA ${path}`);
+        res = await oada.put({
+          tree: isoblueDataTree,
+          path,
+          data: JSON.parse(JSON.stringify({data: data[path].locations})),
```

Comment:
I think `JSON.parse(JSON.stringify(thing))` is a no-op unless:

1. You expect `thing` to change between the call of `oda.put` and it completin
2. `thing` has no JSON serializable things in it that is breaking `@oada/client` (if so, please report the bug to @oada/client)

Otherwise, I think just:

```suggestion
          data: {data: data[path].locations},
```

should work

Reply:
I get the following error when using your suggestion (which is how the code was originally written before the update from oada-client~0.3 to 1.0.2):
```
$ tsc
src/oada_upload.ts(320,11): error TS2322: Type '{ data: { [key: string]: Pick<Location, string | number>; }; }' is not assignable to type 'Json'.
  Types of property 'data' are incompatible.
    Type '{ [key: string]: Pick<V1, string | number>; }' is not assignable to type 'string | number | boolean | Json[] | { [prop: string]: Json; } | null | undefined'.
      Type '{ [key: string]: Pick<V1, string | number>; }' is not assignable to type 'string'.
```

My limited understanding is that the `data` property does not accept the object(?) we created but will accept a JSON object (which is what our object represents). `JSON.stringify`, however, will accept our object and put it into string form which `JSON.parse` can convert into a JSON object.

Possible we could coerce/cast our object into a JSON object directly but I did not think of that at the time

Reply:
Bah, that damn JSON type looked cool, but it is being a pain... Maybe just a 

`data: {data: data[path].locations} as JSON`

you might have to import JSON ... hopefully @oada/client exposes it.

If not, just leave it alone. the `as` method has zero runtime overhead though.

Reply:
This is odd...
```
Type 'JSON' is not assignable to type 'Json'.
  Type 'JSON' is not assignable to type '{ [prop: string]: Json; }'.
    Index signature is missing in type 'JSON'.
```

If you have an idea let me know, otherwise will tackle it after the concurrency stuff

Reply:
Where are you importing the type `JSON` from? Looks like I should have said

 `data: {data: data[path].locations} as Json`

Notice the case differences.